### PR TITLE
db: switch compaction concurrency signal from level score to L0 read-amp

### PR DESCRIPTION
**Pull request by petermattis**
Change the signal used to adjust compaction concurrency from the level
compaction score to L0 read-amp. The level compaction score worked
reasonably for small data sizes, but had the deficiency that the
compaction score tends to decrease over time as the DB size
increases. The level score is based on level size and target level
size. The speed at which this score can change decreases as the level
size gets larger because the input rate of new data is limited. In
contrast, L0 read amp tends to be a reflection of the data input rate
and independent of the DB size (modulo the effect that a larger DB size
tends to have higher write amplification which slows down compacting out
of L0).

The before and after numbers are slightly deceiving as the recent change
to compaction scoring hurt the `kv0/enc=false/nodes=3/size=64kb`,
`kv95/enc=false/nodes=3/size=4kb` and `kv95/enc=false/nodes=3/size=64kb`
workloads for the 10m duration they run at by default. All we're seeing
on those workloads is recovery back to their previous values. The
improvement on `kv0/enc=false/nodes=3/size=4kb` is real and is due to
the smaller number of concurrent compactions.

```
name                              old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3              21.4k ± 4%   21.3k ± 3%     ~     (p=1.000 n=10+9)
kv0/enc=false/nodes=3/size=4kb     3.00k ± 1%   3.66k ± 1%  +22.24%  (p=0.000 n=10+10)
kv0/enc=false/nodes=3/size=64kb      218 ± 1%     256 ± 2%  +17.38%  (p=0.000 n=10+10)
kv95/enc=false/nodes=3             46.7k ± 4%   45.8k ± 2%   -1.96%  (p=0.019 n=10+10)
kv95/enc=false/nodes=3/size=4kb    37.1k ± 2%   37.1k ± 3%     ~     (p=0.912 n=10+10)
kv95/enc=false/nodes=3/size=64kb   4.44k ± 1%   5.19k ± 1%  +16.98%  (p=0.000 n=10+8)
ycsb/A/nodes=3                     22.5k ± 2%   22.5k ± 2%     ~     (p=0.579 n=10+10)
ycsb/B/nodes=3                     40.8k ± 3%   41.4k ± 5%     ~     (p=0.218 n=10+10)
ycsb/C/nodes=3                     56.7k ± 2%   56.3k ± 1%     ~     (p=0.356 n=10+9)
ycsb/D/nodes=3                     44.6k ± 2%   45.3k ± 1%   +1.49%  (p=0.003 n=10+8)
ycsb/E/nodes=3                     6.11k ± 3%   6.03k ± 2%   -1.35%  (p=0.013 n=9+10)
ycsb/F/nodes=3                     9.73k ± 5%  10.13k ± 5%   +4.07%  (p=0.009 n=10+10)
```

### CODE REVIEW

**Comment by petermattis** [comment]
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/769)

**Comment by petermattis** [comment]
Added benchmark numbers. See PR message.

**Comment by petermattis** [comment]
TFTR!

**Comment by itsbilal** [comment]
___
*[compaction_picker.go, line 529 at r2](https://reviewable.io/reviews/cockroachdb/pebble/769#-MAYYyl18vmhX12P6wP-:-MAacJjcDS9pPQmiHun0:bropdp3) ([raw file](https://github.com/cockroachdb/pebble/blob/d45d4ad14f2e9a518bcb1fb21d5917fcde6dd35a/compaction_picker.go#L529)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've added `Options.Experimental.L0CompactionConcurrency`.  I've filed https://github.com/cockroachdb/cockroach/issues/50600 to allow specifying arbitrary Pebble options via the CRDB `--store` flag.
</blockquote></details>

In other words, I was referring to only `return nil`-ing if the number of in-progress L0 {base, intra} compactions is more than 1-per-10 sublevels, and we were going to add another L0 compaction. That would effectively limit concurrency in L0 while allowing it for other levels, but I see the point about wanting to reduce compaction concurrency more generally as well, so this makes sense.
